File: code/packages/csharp/directed-graph/tests/CodingAdventures.DirectedGraph.Tests/DirectedGraphTests.cs
using CodingAdventures.DirectedGraph;

namespace CodingAdventures.DirectedGraph.Tests;

public class DirectedGraphTests
{
    [Fact]
    public void TopologicalSort_OrdersDependencies()
    {
        var graph = new Graph();
        graph.AddEdge("A", "B");
        graph.AddEdge("B", "C");

        Assert.Equal(["A", "B", "C"], graph.TopologicalSort());
    }

    [Fact]
    public void IndependentGroups_LayersDiamond()
    {
        var graph = new Graph();
        graph.AddEdge("A", "B");
        graph.AddEdge("A", "C");
        graph.AddEdge("B", "D");
        graph.AddEdge("C", "D");

        var groups = graph.IndependentGroups();
        Assert.Equal(["A"], groups[0]);
        Assert.Equal(["B", "C"], groups[1]);
        Assert.Equal(["D"], groups[2]);
    }
}

File: code/packages/fsharp/directed-graph/tests/CodingAdventures.DirectedGraph.Tests/DirectedGraphTests.fs
namespace CodingAdventures.DirectedGraph.FSharp.Tests

open CodingAdventures.DirectedGraph.FSharp
open Xunit

module DirectedGraphTests =
    [<Fact>]
    let ``topological sort works`` () =
        let graph =
            DirectedGraph.createDefault()
            |> DirectedGraph.addEdge "A" "B"
            |> DirectedGraph.addEdge "B" "C"

        Assert.Equal<string list>(["A"; "B"; "C"], DirectedGraph.topologicalSort graph)

    [<Fact>]
    let ``graph exposes successors predecessors and affected nodes`` () =
        let graph = Graph()
        graph.AddEdge("logic-gates", "arithmetic")
        graph.AddEdge("arithmetic", "cpu")
        graph.AddEdge("logic-gates", "alu")

        Assert.Equal<string list>(["alu"; "arithmetic"], graph.Successors("logic-gates"))
        Assert.Equal<string list>(["logic-gates"], graph.Predecessors("arithmetic"))
        Assert.Equal<string list>(["logic-gates"; "alu"; "arithmetic"; "cpu"], graph.AffectedNodes([ "logic-gates" ]))

    [<Fact>]
    let ``cycles surface a helpful error`` () =
        let graph = Graph()
        graph.AddEdge("A", "B")
        graph.AddEdge("B", "C")
        graph.AddEdge("C", "A")

        let error = Assert.Throws<CycleError>(fun () -> graph.TopologicalSort() |> ignore)
        Assert.Contains("A", error.Cycle)
